packages/prop-controllers/src/table/table.ts

```diff
@@ -0,0 +1,130 @@
+import { z } from 'zod'
+import {
+  ControlDataTypeKey,
+  CopyContext,
+  Options,
+  Types,
+} from '../prop-controllers'
+import { P, match } from 'ts-pattern'
+
+const tableDataSchema = z.string()
+
+export type TableData = z.infer<typeof tableDataSchema>
+
+const tablePropControllerDataV0Schema = tableDataSchema
+
+export type TablePropControllerDataV0 = z.infer<
+  typeof tablePropControllerDataV0Schema
+>
+
+export const TablePropControllerDataV1Type = 'prop-controllers::table::v1'
+
+const tablePropControllerDataV1Schema = z.object({
+  [ControlDataTypeKey]: z.literal(TablePropControllerDataV1Type),
+  value: tableDataSchema,
+})
+
+export type TablePropControllerDataV1 = z.infer<
+  typeof tablePropControllerDataV1Schema
+>
+
+export const tablePropControllerDataSchema = z.union([
+  tablePropControllerDataV0Schema,
+  tablePropControllerDataV1Schema,
+])
+
+export type TablePropControllerData = z.infer<
+  typeof tablePropControllerDataSchema
+>
+
+export type TableOptions = Options<{
+  preset?: TablePropControllerData
+}>
+
+type TableDescriptorV0<_T = TablePropControllerDataV0> = {
+  type: typeof Types.Table
+  options: TableOptions
+}
+
+type TableDescriptorV1<
+  _T = TablePropControllerData,
+  U extends TableOptions = TableOptions,
+> = {
+  type: typeof Types.Table
+  version: 1
+  options: U
+}
+
+export type TableDescriptor<_T = TablePropControllerData> =
+  | TableDescriptorV0
+  | TableDescriptorV1
+
+export type ResolveTablePropControllerValue<T extends TableDescriptor> =
+  T extends TableDescriptor ? TableData | undefined : never
+
+/**
+ * @deprecated Imports from @makeswift/prop-controllers are deprecated. Use
+ * @makeswift/runtime/controls instead.
+ */
+export function Table(options: TableOptions = {}): TableDescriptorV1 {
+  return { type: Types.Table, version: 1, options }
+}
+
+export function getTablePropControllerDataTableData(
+  data: TablePropControllerData | undefined,
+): TableData | undefined {
+  return match(data)
+    .with(
+      { [ControlDataTypeKey]: TablePropControllerDataV1Type },
+      (v1) => v1.value,
+    )
+    .otherwise((v0) => v0)
+}
+
+export function createTablePropControllerDataFromTableData(
```

Comment:
I think the name here doesn't help much. It's all table data, instead this should probably be `createTablePropControllerDataFromTableId`. Or better yet, it would be nice if this stuff was organized under a namespace or class so we could do `TablePropControllerData.fromId(tableId)`.

Reply:
I agree. Updated to `TableId`.

I have been thinking about the namespace or class since some time ago too 😢